### PR TITLE
Add a Toggle for Low pt Studies + Warnings Printout Toggle

### PR DESCRIPTION
**Pull request by GNiendorf**
Adds a -w option to print out warning statements if they trigger, and a -P option that you can use to specify the pt cut in the standalone version (e.g. `-P 0.6` for 0.6 GeV cut)

### CODE REVIEW

**Comment by GNiendorf** [comment]
@slava77 is there anything wrong with this approach? You can specify the pt cut now by saying 
`sdl_make_tracklooper -mc -P 0.4` for example. It all seems to run fine.

**Comment by VourMa** [comment]
> is there anything wrong with this approach?

Have you tried this in the CMSSW setup? When I try to run, I get:
```
In file included from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Hit.h:4,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Event.h:4,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/LST.h:15,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/RecoTracker/LST/plugins/alpaka/LSTProducer.cc:19:
/home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Constants.h:150:54: error: 'PT_CUT' was not declared in this scope
  150 |     ALPAKA_STATIC_ACC_MEM_GLOBAL const float ptCut = PT_CUT;
      |                                                      ^~~~~~
In file included from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Hit.h:4,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Event.h:4,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/LST.h:15,
                 from /home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/RecoTracker/LST/plugins/alpaka/LSTProducer.cc:19:
/home/users/evourlio/LSTinCMSSW/cgpu-1/CMSSW_13_3_0_pre3/src/../../../TrackLooper/SDL/Constants.h:150:54: error: 'PT_CUT' was not declared in this scope
  150 |     ALPAKA_STATIC_ACC_MEM_GLOBAL const float ptCut = PT_CUT;
      |                                                      ^~~~~~
gmake: *** [config/SCRAM/GMake/Makefile.rules:1793: tmp/el8_amd64_gcc11/src/RecoTracker/LST/plugins/RecoTrackerLSTPluginsPortableCpuSerialSync/alpaka/LSTProducer.cc.o] Error 1
```

Unless this is not yet ready to be run (I though it was, at least at the level I was testing it).

**Comment by GNiendorf** [comment]
I think that can be sort of fixed with a simple ifdef in the config file, i.e. if PT_CUT is not defined then it just goes to a default value of 0.8. I'll look into it more though, not sure if there is a better fix that allows you to select any PT_CUT in CMSSW too.

**Comment by GNiendorf** [comment]
Marked this PR as ready for review. This is just the addition of two toggles, so there are no changes to timing/performance when the two toggles are switched off:

This PR Timing:
Screenshot 2023-11-06 at 11 07 41 PM

This PR: [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR351/mywork_b34c79-PU200/summary/)
Comparison: [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR351/mywork_b34c79-PU200_b34c79-PU200/summary/)

**Comment by VourMa** [comment]
Quick question: Is this meant not to change the pT cuts for the CMSSW setup? I don't see any of changes in the `LST.cc/h`, which are the equivalents of the `trkCore.cc`.

**Comment by GNiendorf** [comment]
Yes, this is just for the standalone version for now. I don't see a reason at this stage to add this to the CMSSW setup, and based on what you said at the last meeting there is an issue with lowering the pt cut which would require more work to sort out.

**Comment by ariostas** [comment]
When I run it with the `-w` option I'm getting the following error. Do you know why this might be happening? It seems to be unrelated to this PR, though. As far as I can tell, this might happen with the DNN, because if you disable it then it works fine.
```
ERROR
 hittype: 0
 trk.pix_simHitIdx().size(): 162652
 trk.ph2_simHitIdx().size(): 213716
162652 0
206907 0
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 206907) >= this->size() (which is 162652)
```

**Comment by GNiendorf** [comment]
@ariostas Can you give your full compile command? I don't see this on my end.

**Comment by ariostas** [comment]
@GNiendorf I'm just using `sdl_run -f "-mcw" -s PU200 -n -1 -t test -d`

File: bin/sdl_run [hunk 1]
```diff
@@ -53,7 +53,7 @@ if [ -z ${TAG} ]; then usage; fi
 
 # If it will not be compiled, then make sure that it was precompiled
 if ${PRECOMPILED}; then
-  if [ ! -f ${TRACKLOOPERDIR}/bin/sdl ] || ( [ ! -f ${TRACKLOOPERDIR}/SDL/libsdl_gpu.so ] && [ ! -f ${TRACKLOOPERDIR}/SDL/libsdl_cpu.so ] ); then
+  if [ ! -f ${TRACKLOOPERDIR}/bin/sdl ] || ( [ ! -f ${TRACKLOOPERDIR}/SDL/libsdl_cuda.so ] && [ ! -f ${TRACKLOOPERDIR}/SDL/libsdl_cpu.so ] ); then
```
Comment (GNiendorf) [inline bin/sdl_run:56]: Minor fix for a previous PR, requested by @ariostas

**Comment by VourMa** [comment]
> `sdl_run -f "-mcw" -s PU200 -n -1 -t test -d`

What does the code do when you add the `-w` flag with no number assigned to it?

**Comment by ariostas** [comment]
> What does the code do when you add the -w flag with no number assigned to it?

There's two different `-w` flags: one for `sdl_make_tracklooper` and one for `sdl`. The one I'm using above is for `sdl_make_tracklooper` and it doesn't take any argument.

**Comment by GNiendorf** [comment]
> @GNiendorf I'm just using `sdl_run -f "-mcw" -s PU200 -n -1 -t test -d`

It seems like this error only occurs when the write_ntuple variable is set to 1. I'll have to look into it more, thank you for pointing this out. Not sure if it is a problem with this PR yet.

**Review by ariostas — APPROVED** [review]
All the changes look good. I tested it and everything works great (aside from the bug with the warnings). Thank you, Gavin!

**Pull request merged**